Bottom: ae87dc7e0a0a1f0db6e3d55a03a816a745504021
Top:    49124e8b905c7ea50a3bdb9a8fad1e035e4b0260
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-02-28 03:03:43 +0000

Refresh of attempt-to-cache-emission-of

---

diff --git a/src/hid/step/object3d_step.c b/src/hid/step/object3d_step.c
index 7006343..f8d3068 100644
--- a/src/hid/step/object3d_step.c
+++ b/src/hid/step/object3d_step.c
@@ -52,6 +52,28 @@ presentation_style_assignments_from_appearance (step_file *step, appearance *app
   return psa_list;
 }
 
+static step_id_list
+lookup_or_create_psa (step_file *step, GHashTable *hash, appearance *appear)
+{
+  step_id_list psa_list;
+
+  if (hash != NULL)
+    psa_list = g_hash_table_lookup (hash, appear);
+
+  if (hash == NULL || psa_list == NULL) {
+    psa_list = presentation_style_assignments_from_appearance (step, appear);
+
+    if (hash != NULL)
+      g_hash_table_insert (hash, appear, psa_list);
+  }
+
+  /* The caller will free the list, so we need to make a copy if the hash table exists */
+  if (hash != NULL)
+    psa_list = g_list_copy (psa_list);
+
+  return psa_list;
+}
+
 static step_file *
 start_ap214_file (const char *filename)
 {
@@ -305,7 +327,7 @@ object3d_to_step_body_fragment (step_file *step,
 #if 1
   /* Body style */
   /* XXX: THERE MUST BE A BODY STYLE, CERTAINLY IF WE WANT TO OVER RIDE FACE COLOURS */
-  brep_style_identifier = step_styled_item (step, "NONE", presentation_style_assignments_from_appearance (step, object->appear), brep_identifier);
+  brep_style_identifier = step_styled_item (step, "NONE", lookup_or_create_psa (step, appear_hash, object->appear), brep_identifier);
   step_presentation_layer_assignment (step, "1", "Layer 1", make_step_id_list (1, brep_style_identifier));
 
   *styled_item_identifiers = step_id_list_append (*styled_item_identifiers, brep_style_identifier);
@@ -315,24 +337,9 @@ object3d_to_step_body_fragment (step_file *step,
     face3d *face = face_iter->data;
 
     if (face->appear != NULL) {
-      step_id_list psa_list;
-      step_id orsi;
-
-      if (appear_hash != NULL)
-        psa_list = g_hash_table_lookup (appear_hash, face->appear);
-
-      if (appear_hash == NULL || psa_list == NULL) {
-        psa_list = presentation_style_assignments_from_appearance (step, face->appear);
-
-        if (appear_hash != NULL)
-          g_hash_table_insert (appear_hash, face->appear, psa_list);
-      }
-
-      if (appear_hash != NULL)
-        psa_list = g_list_copy (psa_list); /* The list is free'd in the step_over_riding_styled_item call */
-
-      orsi = step_over_riding_styled_item (step, "NONE", psa_list,
-                                           face->face_identifier, brep_style_identifier);
+      step_id orsi = step_over_riding_styled_item (step, "NONE",
+                                                   lookup_or_create_psa (step, appear_hash, face->appear),
+                                                   face->face_identifier, brep_style_identifier);
       *styled_item_identifiers = step_id_list_append (*styled_item_identifiers, orsi);
     }
   }
